### PR TITLE
thrift_proxy: decode passthrough data by protocol parsing only for payload_to_metadata filter

### PR DESCRIPTION
**Pull request by JuniorHsu**
…yload_to_metadata filter

Signed-off-by: kuochunghsu <kuochunghsu@pinterest.com>

Commit Message:
This PR fixes two issues:
(a) the decoder actually `drain` the buffer so the router couldn't get the data to send.
we copy to buffer to parse for now as we're waiting for https://github.com/envoyproxy/envoy/issues/23901

(b) passthrough data is not the whole thrift request, 
therefore we introduce `DecoderStateMachine::runPassthroughData` to decode the protocol part only.

Additional Description:
Risk Level: low
Testing: unit test
Docs Changes: filter rst

### CODE REVIEW

File: source/extensions/filters/network/thrift_proxy/decoder.h [hunk 1]
```diff
@@ -88,7 +88,14 @@ class DecoderStateMachine : public Logger::Loggable<Logger::Id::thrift> {
   ProtocolState currentState() const { return state_; }
 
   /**
-   * Set the current state. Used for testing only.
+   * Consumes whole passthrough data without the body start.
```
Comment (zuercher) [inline source/extensions/filters/network/thrift_proxy/decoder.h:91]: nit: I think body here is confusing. Maybe "message start" instead of "body start"?

**Review by zuercher — APPROVED** [review]
One minor comment on the comment, but otherwise looks good.

File: source/extensions/filters/network/thrift_proxy/decoder.h (hunk 1 above)
Comment (JuniorHsu) [reply source/extensions/filters/network/thrift_proxy/decoder.h]: nice suggestion. i addressed this. thanks!

File: docs/root/configuration/other_protocols/thrift_filters/payload_to_metadata_filter.rst [hunk 2]
```diff
@@ -62,7 +62,8 @@ This filter is designed to support payload passthrough. Performing payload to me
 can do deserialization once, and pass the metadata to other filters. This means that load balancing
 decisions, consumed from log and routing could all use payload information with a single parse.
 Also notably performing the parsing in payload passthrough buffer will mean deserialization once
-and not re-serializing, which is the most performant outcome.
+and not re-serializing, which is the most performant outcome. Currently there is an redundant buffer
```
Comment (rgs1) [inline docs/root/configuration/other_protocols/thrift_filters/payload_to_metadata_filter.rst:65]: nit: a redundant

File: test/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/payload_to_metadata_filter_test.cc [hunk 3]
```diff
@@ -75,8 +90,12 @@ class PayloadToMetadataTest : public testing::Test {
   //     f10: set
   //   }
   // }
-  void writeMessage(MessageMetadata& metadata, Buffer::Instance& buffer,
-                    std::string string_value = "two") {
+  void writeMessage(std::string string_value = "two") {
```
Comment (rgs1) [inline test/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/payload_to_metadata_filter_test.cc:93]: thank you for cleaning this up

**Review by rgs1 — APPROVED** [review]
LGTM, just a nit (but you can ignore) too. Thanks!

**Pull request merged**